Automattic/wp-calypso
# Improve runtime performance for `InfiniteList`

Re-submission of #26651 from a branch inside the repo.

#### Changes proposed in this Pull Request

`InfiniteList` currently runs layout synchronously during scrolling. This leads to poor runtime performance:

![image](https://user-images.githubusercontent.com/409615/44003293-53fa243a-9e48-11e8-9c27-a3dcb4f1ad74.png)

This PR delays that work until it's not expensive to run layout, improving things considerably:

![image](https://user-images.githubusercontent.com/409615/44003328-d4c0c20e-9e48-11e8-8e12-74b2aa22bb1d.png)

The remaining long frames are shorter. They seem to be related to React work upon new item load and layout calculations in other scripts (e.g. `w.js`).

This PR also rewrites `InfiniteList` to avoid deprecated React lifecycle methods.

#### Testing instructions

For validating the changes, it's mostly about firing up a local dev instance and navigating to a page with `InfiniteList`, such as the media library or the reader. To measure the performance improvements for you, I'd recommend firing up the Performance tab in Chrome Dev Tools, enabling CPU throttling to 4x or 6x, hitting record, and scrolling the page. Once you stop recording, you should see a chart similar to the above, among other things. You'll want to verify that there are fewer long frames (the ones that are marked in red at the top) and/or the ones that remain are shorter. Let me know if you have any questions!

As for understanding the changes, they're split into two commits:
- The first one simply deals with some housecleaning, by updating the components I'm touching to replace deprecated methods and practices. Since I'm in there, I might as well clean things up :)
- The second one uses the `after-layout-flush` library (in the repo) to schedule DOM updates at a better time, to avoid introducing delays when the browser is trying to produce visual output for the scroll that the user is performing. The [readme for the library](https://github.com/Automattic/wp-calypso/blob/59bdfeeb97eda4266ad39410cb0a074d2c88dbc8/client/lib/after-layout-flush/README.md) offers a good explanation.
